g2p/transducer/__init__.py

```diff
@@ -139,19 +143,164 @@ def tiers(self, value):
             if edge[1] is None:
-                edges[i] = [self._input_nodes[edge[0]][1], None]
+                out_edges.append((self._input_nodes[edge[0]][1], None))
             else:
-                edges[i] = [
-                    self._input_nodes[edge[0]][1],
-                    self._output_nodes[edge[1]][1],
-                ]
-        return edges
+                out_edges.append(
+                    (
+                        self._input_nodes[edge[0]][1],
+                        self._output_nodes[edge[1]][1],
+                    )
+                )
+        return out_edges
 
-    def as_dict(self):
+    def alignments(self) -> List[Tuple[int, int]]:
+        """Return list of alignments (input node index, output node index) for
+        the full transduction.  This *is* the same between TransductionGraph
+        and CompositeTransductionGraph."""
+        return self._edges
+
+    def substring_alignments(  # noqa: C901
+        self, alignments=None
+    ) -> List[Tuple[str, str]]:
+        """Alignments of input to output substrings for this graph.
+
+        As opposed to `pretty_edges`, this method will return the
+        one-to-many or many-to-one alignments produced by a
+        conversion.  It is also the same between TransductionGraph and
+        CompositeTransductionGraph.  For example, assuming this input
+        and output with these edges::
+
+            ABCDEFF
+            aabbcdef
+            [(0, 0), (0, 1), (1, 2), (1, 3), (2, 4), (3, 5),
+             (4, 6), (5, 7), (6, 7)]
+
+        it should return::
+
+            [('A', 'aa'), ('B', 'bb'), ('C', c'), ('D', 'd'),
+             ('E', 'e'), ('FF', 'f')]
+
+        In the case of reorderings, the substrings returned correspond
+        to the minimal *monotonic* alignments, that is, the minimal
+        subsequences which preserve ordering between the input and
+        output.  So, for example given these edges::
+
+            ABCDEF
+            abefcd
+            [(0, 0), (1, 1), (2, 4), (3, 5), (4, 2), (5, 3)]
+
+        it should return::
+
+            [('A', 'a'), ('B', 'b'), ('CDEF', 'efcd')]
+
+        This means, that in the case of truly radical reorderings,
+        there is no "alignment" per se.  This is highly unlikely to
+        occur, given the constrained nature of g2p rules.
+
+        Note that in the case of multi-character outputs such as in
+        ARPABET, the output substrings *do not necessarily* correspond
+        to tokens, and spaces are treated as any other character in
+        the alignment (thus, alignments of (' ', ' ') are possible).
+        If you wish to reconstruct the input it is your resonsibility
+        to collect all the alignments corresponding to an output token
+        (this should not be too hard to do).
+
+        """
+        if alignments is None:
+            alignments = self.alignments()
+
+        def find_monotonic_segments(alignments):
```

Comment:
Ideally I'd split out those functions to stop pylint from complaining but I don't think they are general-purpose enough!

It feels like there is a dynamic programming algorithm to do this but actually I think that just sorting it twice and merging the disagreements is a pretty good and efficient way to do it (since they are already mostly sorted and it's a stable sort)